packages/sdk/src/react/jwtProvider.tsx

```diff
@@ -0,0 +1,171 @@
+  authInfo: null,
+  loading: false,
+  getJwtFromConsentPage: jwtContextNotInitialized,
+  logWithJwt: jwtContextNotInitialized,
+  logOut: jwtContextNotInitialized,
+});
+
+export function useJwtContext(): JwtContextType {
+  return useContext(JwtContext);
+}
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | null>;
+  setItem: (key: string, value: string) => Promise<void>;
+  removeItem: (key: string) => Promise<void>;
+}
+
+interface JwtProviderProps {
+  children: ReactNode;
+  appId: string;
+  storage?: AsyncStorage;
+  storageKeyBuilder?: (appId: string) => string;
+}
+
+export const JwtProvider: React.FC<JwtProviderProps> = ({
+  children,
+  appId,
+  storage = localStorage,
+  storageKeyBuilder = (appId) => `${appId}-jwt`,
+}) => {
+  const appJwtKey = storageKeyBuilder(appId);
+  const vincentWebAppClient = useVincentWebAppClient(appId);
+  const [authInfo, setAuthInfo] = useState<AuthInfo | null>(null);
+  const [loading, setLoading] = useState(true);
+
+  const logOut = useCallback(async () => {
+    try {
+      setLoading(true);
+      setAuthInfo(null);
+      await storage.removeItem(appJwtKey);
+    } finally {
+      setLoading(false);
+    }
+  }, [appJwtKey, storage]);
+
+  const getJwtFromConsentPage = useCallback(
+    (redirectUri: string) => {
+      // Redirect to Vincent Auth consent page with appId and version
+      vincentWebAppClient.redirectToConsentPage({
+        // consentPageUrl: `http://localhost:5173/`,
+        redirectUri,
+      });
+    },
+    [vincentWebAppClient]
+  );
+
+  const logWithJwt = useCallback(async () => {
+    try {
+      setLoading(true);
+
+      const existingJwtStr = await storage.getItem(appJwtKey);
+      const didJustLogin = vincentWebAppClient.isLogin();
+
+      if (didJustLogin) {
+        try {
+          const jwtResult = vincentWebAppClient.decodeVincentLoginJWT(window.location.origin);
+
+          if (jwtResult) {
+            const { decodedJWT, jwtStr } = jwtResult;
+
+            await storage.setItem(appJwtKey, jwtStr);
+            vincentWebAppClient.removeLoginJWTFromURI();
+            setAuthInfo({
+              jwt: jwtStr,
+              pkp: decodedJWT.payload.pkp,
+            });
+            return;
+          } else {
+            await logOut();
+            return;
+          }
+        } catch (e) {
+          console.error('Error decoding JWT:', e);
+          await logOut();
+          return;
+        }
+      }
+
+      if (existingJwtStr) {
+        try {
+          const decodedJWT = verify(existingJwtStr, window.location.origin);
+
+          setAuthInfo({
+            jwt: existingJwtStr,
+            pkp: decodedJWT.payload.pkp,
+          });
+        } catch (error: unknown) {
+          console.error(
+            `Error verifying existing JWT. Need to relogin: ${(error as Error).message}`
+          );
+          await logOut();
+        }
+      }
+    } finally {
+      setLoading(false);
```

Comment:
After reading through this, I think that the _entire_ loading state isn't needed if we just use sync localstorage calls

Reply:
Exactly. `loading` is only needed because of the async storage support